### PR TITLE
feat(models): add UMA (fairchem-core) interatomic-potential wrapper

### PR DESCRIPTION
**Pull request by dallasfoster**
# ALCHEMI Toolkit Pull Request

## Description

Add UMAWrapper, a BaseModelMixin-compatible wrapper around fairchem-core's
UMA (Universal Models for Atoms) MLIPPredictUnit, so UMA foundation models can
drive nvalchemi dynamics and inference. UMA is multi-task: one checkpoint
(uma-s-1p1 / uma-s-1p2 / uma-m-1p1) ships heads for OMol, OMat, OC20, ODAC,
and OMC; the wrapper pins a single task at construction (one-wrapper-one-model,
drive nvalchemi dynamics and inference. UMA is multi-task: one checkpoint
(uma-s-1p1 / uma-s-1p2 / uma-m-1p1) ships heads for OMol, OMat, OC20, ODAC,
and OMC; the wrapper pins a single task at construction (one-wrapper-one-model,
matching MACEWrapper). The conversion is tensor-native (no ASE round trip),
energy is the differentiable primitive, and forces/stress come from autograd.

## Type of Change

- [x] New feature (non-breaking change that adds functionality)

## Changes Made

- nvalchemi/models/uma.py: UMAWrapper (task-aware model_config, adapt_input/
  adapt_output, compute_embeddings, from_checkpoint). from_checkpoint exposes
  fairchem's native `inference_settings` (incl. "turbo"); forward routes the
  one-time lazy-init/MoLE-merge through CPU input to dodge a fairchem device-
  placement bug under turbo on GPU-resident first batches.
- nvalchemi/_optional.py + nvalchemi/models/__init__.py: register/export UMA.
- pyproject.toml: `uma` extra (fairchem-core>=2.0.0); declare uma conflicting
  with mace (e3nn pin) and cu12/cu13 (fairchem torch<2.9 vs toolkit-ops
  torch>=2.11); pin setuptools<81 for fairchem's torchtnt. uv.lock regenerated.
- examples/advanced/09_uma_nve.py: NVE/NVT/NPT MD example driven by built-in
  LoggingHook + EnergyDriftMonitorHook; NPT exercises the stress path; turbo
  selectable via inference_settings.
- test/models/test_uma.py: consolidated suite — structural (mock), forward
  equivalence vs FAIRChemCalculator, charged-input response, NVE drift (@slow),
  turbo/compile device path (@slow, CUDA).
- docs: userguide/models.md (supported-models table + UMA usage / HF-token /
  torch-environment notes), modules/models.rst, models/index.md, examples README.
- .github/workflows/ci.yml: install a dedicated .venv-uma and run the UMA tests
  from it (gated on UMA-file changes or full runs); optional HF_TOKEN secret.

## Testing

- [x] Unit tests pass locally (UMA suite: 36 passed with --slow against
      uma-s-1p1 on CUDA; 33 passed / 3 slow-skipped without --slow)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality

Equivalence matches FAIRChemCalculator to 1e-4 (OMol energy/forces, OMat
energy/forces/stress); charged OMol runs match the calculator at charge -1;
small (uma-s-1p1/1p2) and medium (uma-m-1p1) checkpoints load and run.

## Additional Notes

UMA's deps conflict with the mace/cuXX stack, so it must be installed in its
own environment (`uv sync --extra uma`); it brings its own CUDA-enabled torch
(2.8, cu12.8) and does not use the cuXX GPU stack. Checkpoints are gated on
HuggingFace (facebook/UMA) — CI structural tests run without a token; the
checkpoint-based tests skip unless an HF_TOKEN secret is provided.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by dallasfoster** [comment]
/ok to test [8344178](https://github.com/NVIDIA/nvalchemi-toolkit/pull/117/commits/83441789933fdde24811c5ff3fce28ce40142260)

File: nvalchemi/models/uma.py [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+    NeighborConfig,
+    NeighborListFormat,
+)
+
+__all__ = ["UMAWrapper"]
+
+
+# Task names accepted by fairchem / UMA. Kept as a module-level tuple
+# so the wrapper can validate at construction and tests can iterate.
+_UMA_TASKS: tuple[str, ...] = ("omol", "omat", "oc20", "odac", "omc")
```
Comment (laserkelvin) [inline nvalchemi/models/uma.py]: Why not a set as well?

File: nvalchemi/models/uma.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+
+    Attributes
+    ----------
+    model_config : ModelConfig
+        Task-dependent outputs + autograd + neighbor config.
+    task_name : str
+        The UMA task this wrapper is pinned to.
+    """
+
+    def __init__(self, predict_unit: Any, task_name: str = "omol") -> None:
```
Comment (laserkelvin) [inline nvalchemi/models/uma.py]: Can we tighten up the `Any`?

File: nvalchemi/models/uma.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+            Optional overrides forwarded to fairchem's inference-settings
+            builder.
+        """
+        import os as _os
+
+        from fairchem.core.calculate import pretrained_mlip  # noqa: PLC0415
+        from fairchem.core.units.mlip_unit import load_predict_unit  # noqa: PLC0415
+
+        if isinstance(device, torch.device):
+            device = device.type
```
[the remark this answers is not in the document]
Comment (dallasfoster) [reply nvalchemi/models/uma.py:284]: I verified fairchem's get_predict_unit signature is device: Literal['cuda','cpu'] and it routes the actual CUDA device via get_device_for_local_rank(). Passing "cuda:1" (what str() produces) is treated as not-"cuda" → falls back to CPU — a worse outcome.

File: nvalchemi/models/uma.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+        ``predict`` moves it back to the model device, so inference stays
+        on-device and later forwards use the input's real device.
+        """
+        fc_data = self.adapt_input(data, **kwargs)
+
+        # Route only the first (lazy-init) forward through CPU input under
+        # turbo — see the docstring for why.
+        settings = getattr(self.predict_unit, "inference_settings", None)
+        if (
+            not getattr(self.predict_unit, "lazy_model_intialized", True)
```
[the remark this answers is not in the document]
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: fairchem genuinely has that misspelling, so our getattr matched today — but Greptile's forward-looking concern (a future fairchem rename would silently disable the workaround) is fair. I removed the dependency on fairchem's private flag entirely: the wrapper now tracks its own one-shot _cpu_route_first_forward flag set at construction from inference_settings. More robust and clearer.

File: nvalchemi/models/uma.py [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+
+    # ------------------------------------------------------------------
+    # Construction
+    # ------------------------------------------------------------------
+
+    @classmethod
+    def from_checkpoint(
+        cls,
+        name_or_path: str | Path,
+        task_name: str = "omol",
```
Comment (laserkelvin) [inline nvalchemi/models/uma.py]: Should be typed as `Literal` or point to the `_UMA_TASKS` set/tuple

File: nvalchemi/models/uma.py [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+                name_str,
+                inference_settings=inference_settings,
+                overrides=overrides,
+                device=device,
+            )
+        else:
+            raise ValueError(
+                f"{name_str!r} is neither a registered model name nor a "
+                f"local file path. Known names: "
+                f"{sorted(pretrained_mlip.available_models)[:6]}..."
```
Comment (laserkelvin) [inline nvalchemi/models/uma.py]: Why is it truncated to the 6 items? You can just leave an inline comment maybe

**Comment by dallasfoster** [comment]
/ok to test [00044fa](https://github.com/NVIDIA/nvalchemi-toolkit/pull/117/commits/00044fa9aee6b5910dccb26bd2c54c9773937cca)

File: nvalchemi/models/uma.py [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+        # wants per-system long (B,), so flatten + cast (also handles raw (B,)).
+        charge = getattr(data, "charge", None)
+        if charge is None:
+            charge = torch.zeros(n_systems, dtype=torch.long, device=device)
+        else:
+            charge = charge.to(torch.long).reshape(n_systems)
+
+        spin = getattr(data, "spin", None)
+        if spin is None:
+            spin = torch.zeros(n_systems, dtype=torch.long, device=device)
```
Comment (laserkelvin) [inline nvalchemi/models/uma.py]: This should not be zero. A sensible default could be 1, but it really depends on the system's unpaired electrons.

File: examples/advanced/09_uma_nve.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        TASK = "omat"
+
+# %%
+# Build the system
+# ----------------
+# Helpers turn an ASE ``Atoms`` object into an nvalchemi ``Batch`` with
+# Maxwell-Boltzmann velocities at ``TEMPERATURE_K`` and zero net momentum.
+# Periodic systems carry ``cell`` and ``pbc``.
+
+_PROPANE_POSITIONS = np.array(
```
Comment (laserkelvin) [inline examples/advanced/09_uma_nve.py:172]: Could we make this example batched? You could duplicate the system and just use modified positions to start the simulations

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Generally looks good to me; the spin default value needs to change though

File: nvalchemi/models/uma.py (hunk 1 above)
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: fixed

File: nvalchemi/models/uma.py (hunk 5 above)
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: done.

File: nvalchemi/models/uma.py (hunk 6 above)
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: fixed.

File: nvalchemi/models/uma.py (hunk 2 above)
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: resolving to "MLIPPredictUnit"

File: nvalchemi/models/uma.py (hunk 7 above)
Comment (dallasfoster) [reply nvalchemi/models/uma.py]: We use spin default = 1 for OMol and 0 for periodic - which comes from fairchem's documentation.

File: examples/advanced/09_uma_nve.py (hunk 8 above)
Comment (dallasfoster) [reply examples/advanced/09_uma_nve.py:172]: done.

**Comment by dallasfoster** [comment]
/ok to test [816b230](https://github.com/NVIDIA/nvalchemi-toolkit/pull/117/commits/816b2309bdb0446fbb5e346fbc02eab69c82cfc8)

**Review by laserkelvin — APPROVED** [review]
